### ISSUE TITLE
Again 'loadRobotModel': std::bad_alloc Error 

### ISSUE THREAD

**Issue by JoNeckel** [UniversalRobots/Universal_Robots_ROS2_Driver#230]
Version: foxy

I had this Issue: https://github.com/UniversalRobots/Universal_Robots_ROS2_Driver/issues/197#issue-1019730249 a few weeks ago, but it was solved with this commit: ca5bc40ac05f207712222a6abcb9e22c1c2d4584 .

Now @TsuunEvo had a similar Error and I reproduced it by rebuilding the driver.

the error occurs when executing these commands:
```
ros2 launch ur_bringup ur_control.launch.py ur_type:=ur5e robot_ip:=yyy.yyy.yyy.yyy use_fake_hardware:=true launch_rviz:=false


ros2 launch ur_bringup ur_moveit.launch.py ur_type:=ur5e robot_ip:="xxx.xxx" use_fake_hardware:=true launch_rviz:=true

```

this is the terminal output:

```
ros2 launch ur_bringup ur_moveit.launch.py ur_type:=ur5e robot_ip:="xxx.xxx" use_fake_hardware:=true launch_rviz:=true
[INFO] [launch]: All log files can be found below /home/jo/.ros/log/2021-10-28-18-57-00-505520-myubuntu-38988
[INFO] [launch]: Default logging verbosity is set to INFO
[INFO] [move_group-1]: process started with pid [38995]
[INFO] [mongo_wrapper_ros.py-2]: process started with pid [38997]
[INFO] [rviz2-3]: process started with pid [38999]
[INFO] [static_transform_publisher-4]: process started with pid [39001]
[static_transform_publisher-4] [INFO] [1635440221.371882197] [static_transform_publisher]: Spinning until killed publishing transform from 'world' to 'base_link'
[move_group-1] [WARN] [1635440221.799056047] [move_group.move_group]: MoveGroup launched without ~default_planning_pipeline specifying the namespace for the default planning pipeline configuration
[move_group-1] [WARN] [1635440221.799131393] [move_group.move_group]: Falling back to using the the move_group node namespace (deprecated behavior).
[move_group-1] Parsing robot urdf xml string.
[move_group-1] [INFO] [1635440221.806328362] [moveit_rdf_loader.rdf_loader]: Loaded robot model in 0.00255315 seconds
[move_group-1] [INFO] [1635440221.806395974] [moveit_robot_model.robot_model]: Loading robot model 'ur'...
[move_group-1] terminate called after throwing an instance of 'std::bad_alloc'
[move_group-1]   what():  std::bad_alloc
[mongo_wrapper_ros.py-2] [INFO] [1635440221.848078191] [mongodb]: /tmp/db does not exist; creating it.
[mongo_wrapper_ros.py-2] [INFO] [1635440221.848985167] [mongodb]: Starting mongodb with db location /tmp/db listening on localhost:33829
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:01.895+0200 I CONTROL  [initandlisten] MongoDB starting : pid=39049 port=33829 dbpath=/tmp/db 64-bit host=myubuntu
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:01.895+0200 I CONTROL  [initandlisten] db version v3.6.8
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:01.895+0200 I CONTROL  [initandlisten] git version: 8e540c0b6db93ce994cc548f000900bdc740f80a
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:01.895+0200 I CONTROL  [initandlisten] OpenSSL version: OpenSSL 1.1.1f  31 Mar 2020
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:01.895+0200 I CONTROL  [initandlisten] allocator: tcmalloc
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:01.895+0200 I CONTROL  [initandlisten] modules: none
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:01.895+0200 I CONTROL  [initandlisten] build environment:
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:01.895+0200 I CONTROL  [initandlisten]     distarch: x86_64
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:01.895+0200 I CONTROL  [initandlisten]     target_arch: x86_64
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:01.895+0200 I CONTROL  [initandlisten] options: { net: { port: 33829 }, storage: { dbPath: "/tmp/db" } }
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:01.895+0200 I STORAGE  [initandlisten] 
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:01.895+0200 I STORAGE  [initandlisten] ** WARNING: Using the XFS filesystem is strongly recommended with the WiredTiger storage engine
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:01.895+0200 I STORAGE  [initandlisten] **          See http://dochub.mongodb.org/core/prodnotes-filesystem
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:01.895+0200 I STORAGE  [initandlisten] wiredtiger_open config: create,cache_size=7438M,session_max=20000,eviction=(threads_min=4,threads_max=4),config_base=false,statistics=(fast),cache_cursors=false,compatibility=(release="3.0",require_max="3.0"),log=(enabled=true,archive=true,path=journal,compressor=snappy),file_manager=(close_idle_time=100000),statistics_log=(wait=0),verbose=(recovery_progress),
[ERROR] [move_group-1]: process has died [pid 38995, exit code -6, cmd '/home/jo/workspace2/ros_ur_driver/install/moveit_ros_move_group/lib/moveit_ros_move_group/move_group --ros-args --params-file /tmp/launch_params_cp1516ff --params-file /tmp/launch_params_5n71vdhp --params-file /tmp/launch_params_n22nlzd8 --params-file /tmp/launch_params_a9tldbh3 --params-file /tmp/launch_params_eu70qffg --params-file /tmp/launch_params_a33px69o --params-file /tmp/launch_params_sl6mfogw'].
[rviz2-3] [INFO] [1635440222.469861788] [rviz2]: Stereo is NOT SUPPORTED
[rviz2-3] [INFO] [1635440222.469966764] [rviz2]: OpenGl version: 4.6 (GLSL 4.6)
[rviz2-3] [INFO] [1635440222.510429606] [rviz2]: Stereo is NOT SUPPORTED
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:02.542+0200 I STORAGE  [initandlisten] WiredTiger message [1635440222:542889][39049:0x7f287f99ff40], txn-recover: Set global recovery timestamp: 0
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:02.606+0200 I CONTROL  [initandlisten] 
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:02.606+0200 I CONTROL  [initandlisten] ** WARNING: Access control is not enabled for the database.
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:02.606+0200 I CONTROL  [initandlisten] **          Read and write access to data and configuration is unrestricted.
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:02.606+0200 I CONTROL  [initandlisten] 
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:02.606+0200 I CONTROL  [initandlisten] ** WARNING: This server is bound to localhost.
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:02.606+0200 I CONTROL  [initandlisten] **          Remote systems will be unable to connect to this server. 
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:02.606+0200 I CONTROL  [initandlisten] **          Start the server with --bind_ip <address> to specify which IP 
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:02.606+0200 I CONTROL  [initandlisten] **          addresses it should serve responses from, or with --bind_ip_all to
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:02.606+0200 I CONTROL  [initandlisten] **          bind to all interfaces. If this behavior is desired, start the
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:02.606+0200 I CONTROL  [initandlisten] **          server with --bind_ip 127.0.0.1 to disable this warning.
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:02.606+0200 I CONTROL  [initandlisten] 
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:02.606+0200 I STORAGE  [initandlisten] createCollection: admin.system.version with provided UUID: 4552777f-823e-4136-b2bb-1aacd8d2735e
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:02.639+0200 I COMMAND  [initandlisten] setting featureCompatibilityVersion to 3.6
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:02.642+0200 I STORAGE  [initandlisten] createCollection: local.startup_log with generated UUID: 3c72c821-80da-4b20-9bba-e83e4e6e8e19
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:02.665+0200 I FTDC     [initandlisten] Initializing full-time diagnostic data capture with directory '/tmp/db/diagnostic.data'
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:02.666+0200 I NETWORK  [initandlisten] waiting for connections on port 33829
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:02.666+0200 I STORAGE  [LogicalSessionCacheRefresh] createCollection: config.system.sessions with generated UUID: 19005eee-9f20-4b16-bbed-95382f026565
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:02.695+0200 I INDEX    [LogicalSessionCacheRefresh] build index on: config.system.sessions properties: { v: 2, key: { lastUse: 1 }, name: "lsidTTLIndex", ns: "config.system.sessions", expireAfterSeconds: 1800 }
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:02.695+0200 I INDEX    [LogicalSessionCacheRefresh] 	 building index using bulk method; build may temporarily use up to 500 megabytes of RAM
[mongo_wrapper_ros.py-2] 2021-10-28T18:57:02.696+0200 I INDEX    [LogicalSessionCacheRefresh] build index done.  scanned 0 total records. 0 secs
[rviz2-3] Warning: class_loader.impl: SEVERE WARNING!!! A namespace collision has occured with plugin factory for class rviz_default_plugins::displays::InteractiveMarkerDisplay. New factory will OVERWRITE existing one. This situation occurs when libraries containing plugins are directly linked against an executable (the one running right now generating this message). Please separate plugins out into their own library or just don't link against the library and use either class_loader::ClassLoader/MultiLibraryClassLoader to open.
[rviz2-3]          at line 253 in /opt/ros/foxy/include/class_loader/class_loader_core.hpp
[rviz2-3] [ERROR] [1635440225.780069617] [moveit_ros_visualization.motion_planning_frame]: Action server: /recognize_objects not available
[rviz2-3] Parsing robot urdf xml string.
[rviz2-3] [INFO] [1635440225.891697207] [moveit_rdf_loader.rdf_loader]: Loaded robot model in 0.00228706 seconds
[rviz2-3] [INFO] [1635440225.892683553] [moveit_robot_model.robot_model]: Loading robot model 'ur'...
[rviz2-3] [ERROR] [1635440225.912147061] [moveit_background_processing.background_processing]: Exception caught while processing action 'loadRobotModel': std::bad_alloc
```

**Comment by makeadl** [comment]
I have also the same problem. I tried also with the old commits but still have the same issue.

**Comment by JamesTRoss1** [comment]
I have the same issue through cloning the foxy branch.

**Comment by hodgespodge** [comment]
I also have the same issue from cloning the foxy branch.

**Comment by JamesTRoss1** [comment]
Has anyone figured out the issue? Would love an update on the issue.

**Comment by JoNeckel** [comment]
It is really strange, sometimes it works and sometimes it doesn't...
I haven't found a solution yet

**Comment by JoNeckel** [comment]
There is a fix now! 
The problem is in the MoveIt source. 
see: https://github.com/ros-planning/moveit2/issues/780
You can build MoveIt from source with the pr branch 782 (`gh pr checkout 782 `or `git remote add foxy_fix https://github.com/JafarAbdi/moveit2.git && git fetch foxy_fix pr-foxy_backports && git checkout pr-foxy_backports` after cloning the MoveIt source)

**Comment by makeadl** [comment]
I still have the same issue after using:

git remote add foxy_fix https://github.com/JafarAbdi/moveit2.git && git fetch foxy_fix pr-foxy_backports && git checkout pr-foxy_backports

my git status looks like:

workspace/ros_ur_driver/src/moveit2$ git status
On branch pr-foxy_backports
Your branch is up to date with 'foxy_fix/pr-foxy_backports'.

nothing to commit, working tree clean

should I remove all these build log and install folder and build it again ?

Note: I did the git remote command inside the moveit2 folder. is that right ?

@JoNeckel

**Comment by JoNeckel** [comment]
@makeadl I did `gh pr checkout 782` inside /workspace/ros_ur_driver/src/moveit2 but it did not work on the first try (i think your git remote should do the same). 
It worked after I removed all moveit directorys inside /opt/ros/foxy/install/ and then rebuilt moveit inside the workspace after also removing the /build /log and /install folders.

Now I don't have any moveit folders inside /opt/ros/foxy/install anymore, i don't know if it was a good idea to remove these folders but it seems to work for now.

**Comment by makeadl** [comment]
did you mean opt/ros/foxy/share? I don't have the install folder in my the opt/ros/foxy folder

**Comment by JoNeckel** [comment]
@makeadl I think it was an install folder, maybe there was something wrong with my filesytem, but maybe I'm wrong and it was the share folder

**Comment by makeadl** [comment]
Okey I got it. 
you must  remove moveit, if you have installed it with binary installation: sudo apt install ros-foxy-moveit.

sudo apt-get purge ros-foxy-moveit. and then checkout the right commit with : 

gh pr checkout 782

or

git remote add foxy_fix https://github.com/JafarAbdi/moveit2.git && git fetch foxy_fix pr-foxy_backports && git checkout pr-foxy_backports

thank you so much

**Comment by sabrinaacardoso** [comment]
I follow all the steps but still have the same problem here. Any help?

**Comment by JoNeckel** [comment]
@sabrinaacardoso 
I would try to remove everything inside ~/workspace/ros_ur_driver/

If you installed the binary version of moveit you have to `apt purge ros-foxy-moveit` it, like makeadl said. 
Then search inside /opt/ros/foxy/ if there are any directories with "moveit" in their name and if so, remove them.
Then inside follow these instructions to build moveit from source: https://moveit.ros.org/install-moveit2/source/
But not inside ws_moveit2/src but instead in ~/workspace/ros_ur_driver/src and after `git clone https://github.com/ros-planning/moveit2.git -b foxy`

You have to do:
`gh pr checkout 782`  (if you have the gh tool)

or

`git remote add foxy_fix https://github.com/JafarAbdi/moveit2.git && git fetch foxy_fix pr-foxy_backports && git checkout pr-foxy_backports`

After that, go on with vcs import and so on..
After building moveit, you can start buildung the Universal_Robots_ROS2_Driver like in the instructions (this should not take much time anymore

**Comment by sabrinaacardoso** [comment]
It worked here!! Thank you so much.

**Comment by JoNeckel** [comment]
This Issue should now be fixed with fd9a562830bd5b3720905cd01dbf97e7591ca31e.
